### ISSUE TITLE
Bug: [explicit-function-return-type] Return type not accepted on global typed function expression

### ISSUE THREAD

**Issue by aaronadamsCA** [typescript-eslint/typescript-eslint#5019]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.0-beta&sourceType=module&code=KYDwDg9gTgLgBAbwL4G4CwAoTATYBjAGwEMpg4BzAiAIyIMUzjgEsA7GYKAMyLzIHU22CAHcGGJky4QIALjgAKAJRwAvAD44ANwjNs6CXCSZjWDCKGiAdNIhrFKjXDwRWAZwgFgVttIUAiAAlgAio4EWgCbH8lAyA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9HAB4z5JrYBmArgUhGgPYFYpwT0ofTwQALhBwUKVihABfIA&tsConfig=N4XyA

### Repro Code

```typescript
export {};

declare global {
  interface Window {
    foo: () => void;
  }
}

window.foo = () => console.info("Hello world");
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/explicit-function-return-type": "error",
  },
};
```

### tsconfig

```jsonc
N/A
```

### Expected Result

No errors, because of default option `allowTypedFunctionExpressions: true`.

### Actual Result

Missing return type on function.

### Versions

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.25.0` |
| `@typescript-eslint/parser`        | `5.25.0` |
| `TypeScript`                       | `4.7.0-beta` |
| `ESLint`                           | `8.15.0` |
| `Env`                              | `web` |

**Comment by bradzacher** [comment]
Assignment expressions are not seen as typed locations. This is done on purpose.

For example this code similarly will error:
```ts
foo = () => 1
```

The intention of the option is not "allow locations where the return type can be inferred by the checker", it is instead "allow locations where the return type is explicitly declared nearby".

https://typescript-eslint.io/rules/explicit-function-return-type#allowtypedfunctionexpressions
